shpc/main/registry/filesystem.py

```diff
@@ -75,20 +75,31 @@ def override_exists(self, tag):
 
 
 class Filesystem(Provider):
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.source = os.path.abspath(self.source)
+    def __init__(self, source):
+        if not self.matches(source):
+            raise ValueError(
+                "Filesystem registry source must exist on the filesystem. Got %"
+                % source
+            )
+        self.source = os.path.abspath(source)
 
     @classmethod
     def matches(cls, source):
         return os.path.exists(source) or source == "."
 
+    def exists(self, name):
+        return os.path.exists(os.path.join(self.source, name))
+
     def iter_modules(self):
+        """
+        yield module names
+        """
+        # Find modules based on container.yaml
         for filename in shpc.utils.recursive_find(self.source, "container.yaml"):
-            module = os.path.dirname(filename).replace(self.source, "").strip(os.sep)
+            module = os.path.dirname(filename)
```

Comment:
This is going to miss the entire module uri (e.g., quay.io/biocontainers/samtools) and only provide "samtools"

Reply:
Oups, sorry

Reply:
Haha no worries! That's what the review is for!

Reply:
Can you explain the issue ? `filename` has indeed changed and is now the relative path of `container.yaml` (included) from `self.source`, but `module` is still the relative directory path that holds `container.yaml`.

Reply:
It's hard to say without seeing how this function is changed - I think what we want here is to be able to get from:

- quay.io/biocontainers/samtools/container.yaml (fullpath or not) to
- quay.io/biocontainers/samtools

Before we were starting with the dirname of that, so:

- /full/path/modules/quay.io/biocontainers/samtools/container.yaml

and deriving

- quay.io/biocontainers/samtools

by replacing the source `/fullpath/modules` + a strip of os.sep

So perhaps a sanity check is to run this and print what you get for your new module - if it's the full `quay.io/biocontainers/samtools` then you are probably OK - it totally depends on what that (potentially now changed) function returns.

And I'm off for tonight - will make some more time tomorrow evening to review further!

Reply:
I can confirm it returns "quay.io/biocontainers/samtools" (and the tests pass, and the various shpc commands I've tried work as well)